src/zimcheck/checks.cpp

```diff
@@ -702,22 +702,100 @@ void test_articles(const zim::Archive& archive, ErrorLogger& reporter, ProgressB
     }
 }
 
-void test_redirect_loop(const zim::Archive& archive, ErrorLogger& reporter) {
-    reporter.infoMsg("[INFO] Checking for redirect loops...");
+namespace
+{
+
+class RedirectionTable
+{
+private: // types
+    enum LoopStatus : uint8_t
+    {
+        UNKNOWN,
+        LOOP,
+        NONLOOP
+    };
+
+public: // functions
+    void addRedirectionEntry(zim::entry_index_type targetEntryIndex)
+    {
+        redirTable.push_back(targetEntryIndex);
+        loopStatus.push_back(LoopStatus::UNKNOWN);
+    }
+
+    void addItem()
+    {
+        redirTable.push_back(redirTable.size());
+        loopStatus.push_back(LoopStatus::NONLOOP);
+    }
 
-    int chained_redirection_limit = 50;
+    size_t size() const { return redirTable.size(); }
 
-    for(auto& entry: archive.iterEfficient())
+    bool isInRedirectionLoop(zim::entry_index_type i)
     {
-        auto current_entry = entry;
-        int redirections_done = 0;
-        while(current_entry.isRedirect() && redirections_done < chained_redirection_limit)
+        if ( loopStatus[i] == UNKNOWN )
         {
-            current_entry = current_entry.getRedirectEntry();
-            redirections_done++;
+            resolveLoopStatus(i);
         }
 
-        if(current_entry.isRedirect()){
+        return loopStatus[i] == LOOP;
+    }
+
+private: // functions
+    LoopStatus detectLoopStatus(zim::entry_index_type i) const
+    {
+        auto i1 = i;
+        auto i2 = i;
+        // Follow redirections until an entry with known loop status
+        // is found.
+        // i2 moves through redirections at twice the speed of i1
+        // if i2 runs over i1 then they are both inside a redirection loop
+        for (bool moveI1 = false ; ; moveI1 = !moveI1)
+        {
+            if ( loopStatus[i2] != LoopStatus::UNKNOWN )
+                return loopStatus[i2];
+
+            i2 = redirTable[i2];
+
+            if ( i2 == i1 )
+                return LoopStatus::LOOP;
+
+            if ( moveI1 )
+                i1 = redirTable[i1];
+        }
+    }
+
+    void resolveLoopStatus(zim::entry_index_type i)
+    {
+        const LoopStatus s = detectLoopStatus(i);
+        for ( ; loopStatus[i] == LoopStatus::UNKNOWN; i = redirTable[i] )
+        {
+            loopStatus[i] = s;
+        }
+    }
+
+private: // data
+    std::deque<zim::entry_index_type> redirTable;
+    std::deque<LoopStatus> loopStatus;
```

Comment:
Because of `std::vector`'s higher memory usage requirements during growth. When you incrementally populate an `std::vector` with `N>>1` elements its high-watermark memory usage is at about 3/2N (~N/2 of previous storage + ~N new storage during the last reallocation). `std::deque` typically being implemented as a vector of dynamically allocated fixed size blocks is devoid of that disadvantage.

Reply:
We could reserve the space once at initialization as we know the number of entries to add

Reply:
Yes, but then we will need to separately track the count of items already added. What are the cons of `std::deque`?

Reply:
> Yes, but then we will need to separately track the count of items already added. 

With `std::vector::reserve` we can pre-allocate the memory for the vector without changing the semantic.
Or I misunderstand your argument...

> What are the cons of std::deque?

The semantic. `std::deque` is a queue. Which means we use to to pass elements in order (add with `push_back`, read with `pop_front`). Here we have the semantic of a growing vector, so we should use a vector.
We may have the hack of using deque for memory allocation improvement. But as we can reserve the memory with
`redirTable.reserve(archive.getAllEntryCount())`, there is no real reason to use it.

Reply:
> With `std::vector::reserve` we can pre-allocate the memory for the vector without changing the semantic.
> Or I misunderstand your argument...

Indeed. Totally forgot about `reserve()`

Reply:
Done